### PR TITLE
added configurator , task manager and task packages

### PR DESCRIPTION
**Pull request by Nikhil-dot-Kumar**
Checklist:

- [x] I have read [contribution guidelines](https://github.com/maplelabs/github-audit/blob/master/CONTRIBUTING.md).
- [x] My pull request has a descriptive title (not a vague title like `Update index.md`)
- [x] I have tested these changes locally on my machine.

##### SUMMARY

1. Added core engine for github-audit 
2. Configurator package creates tasks for correct audit job configs.
3. Task manager schedules tasks created by configurator
4. Task contains task related logic

##### COMPONENT NAME

[Configurator](https://github.com/maplelabs/github-audit/blob/master/docs/design/software_design.md#configurator)
[Task Manger](https://github.com/maplelabs/github-audit/blob/master/docs/design/software_design.md#task-manager)
[Task](https://github.com/maplelabs/github-audit/blob/master/docs/design/software_design.md#task)

##### ADDITIONAL INFORMATION

[Datastore](https://github.com/maplelabs/github-audit/blob/master/docs/design/software_design.md#in-memory-datastore) package in not required from original design , same operations are added in task package. 

```paste below

```

### CODE REVIEW

File: internal/configurator/configurator.go [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+	var err error
+	task := task.Newtask()
+	if auditJob.AccessToken != "" {
+		decodedKey, err = DecodeAccessKey(auditJob.AccessToken)
+		if err != nil {
+			log.Errorf("error[%v] in decoding accessToken for audit job %v", err, auditJob.Name)
+			return task, err
+		}
+	}
+	gp := gitprovider.NewGitProvider(auditJob.RepositoryHost, auditJob.RepositoryOwner, auditJob.RepositoryName, auditJob.Username, decodedKey)
```
Comment (upendrasahu) [inline internal/configurator/configurator.go:65]: Function argument can be auditJob object, you are passing individual atributes?

File: internal/configurator/configurator.go [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+func createTaskParam(auditJob input.AuditJob, targets []input.Target, decodedKey string) task.TaskParams {
+	var taskParam task.TaskParams
+	taskParam.Config = auditJob
+	taskParam.Targets = targets
+	taskParam.DecodeAccessKey = decodedKey
+	return taskParam
+}
+
+// DecodeAccessKey decodes the base64 encoded access key as provided in config.yaml or through environment variable.
+func DecodeAccessKey(key string) (string, error) {
```
Comment (upendrasahu) [inline internal/configurator/configurator.go:88]: Can this function be defined in separate utils to be reused?

Comment (Nikhil-dot-Kumar) [reply internal/configurator/configurator.go:88]: adding a separate utils package for this

File: internal/configurator/configurator.go [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+)
+
+var (
+	log logger.Logger
+	// ErrNoTaskConfigured if there is no audit job configured.
+	ErrNoTaskConfigured = errors.New("no audit task configured")
+)
+
+func init() {
+	log = logger.GetLogger()
```
Comment (upendrasahu) [inline internal/configurator/configurator.go:30]: Make sure logs have info on which modue log is coming from. there are logger which privide that.

File: internal/configurator/configurator.go [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+
+// StartProcessing returns list of tasks with valid credentials which can be scheduled.
+func StartProcessing(config input.Config) ([]*task.Task, error) {
+	log.Info("starting creation of tasks for configured audit jobs")
+	tasks := make([]*task.Task, 0)
+	for _, aj := range config.AuditJobs {
+		task, err := createTask(aj, config.Targets)
+		if err != nil {
+			log.Errorf("error[%v] in creating task for audit job %v", err, aj.Name)
+			continue
```
Comment (upendrasahu) [inline internal/configurator/configurator.go:41]: How are you retrying tasks if few tasks fails and few passes?

File: internal/configurator/configurator.go (hunk 3 above)
Comment (Nikhil-dot-Kumar) [reply internal/configurator/configurator.go:30]: log have the file name from which file it is coming from , should i add for package also?

File: internal/configurator/configurator.go (hunk 4 above)
Comment (Nikhil-dot-Kumar) [reply internal/configurator/configurator.go:41]: nope , if a task creation fails , only adding logs to log file for user to review the task config

File: internal/task/task.go [hunk 5]
```diff
@@ -0,0 +1,314 @@
...
+
+//TODO: add stop function in future if required
+func (t *Task) Stop() error {
+	return nil
+}
+
+// collectAndPublishCommits collects commits and publish them to targets.
+func (t *Task) collectAndPublishCommits(gp gitprovider.GitProvider, pb publisher.Publisher, dp dataprocessor.DataProcessor, ts TaskStats) error {
+	var err error
+	for _, br := range t.Config.Branches {
```
Comment (upendrasahu) [inline internal/task/task.go:232]: Instead of running loop, can you run concurrent go routines.

File: internal/task/task.go [hunk 6]
```diff
@@ -0,0 +1,314 @@
...
+		if err != nil {
+			log.Error(err)
+		}
+	}
+}
+
+// SaveTaskStatsPeriodic runs periodically to save task stats to file system (checkpointing mechanism).
+func SaveTaskStatsPeriodic(ctx context.Context) {
+	// 30 seconds periodic
+	ticker := time.NewTicker(30 * time.Second)
```
Comment (upendrasahu) [inline internal/task/task.go:105]: Add this interval in const

File: internal/task/task.go (hunk 5 above)
Comment (Nikhil-dot-Kumar) [reply internal/task/task.go:232]: sure will add anonymous go func

File: internal/task/task.go [hunk 7]
```diff
@@ -0,0 +1,314 @@
...
+		ts.LastPullRequestNo = 0
+		ts.TaskID = t.ID
+		// saving default task stats for a task.
+		saveTaskStats(t.ID, ts)
+	}
+	gp := gitprovider.NewGitProvider(t.Config.RepositoryHost, t.Config.RepositoryOwner, t.Config.RepositoryName, t.Config.Username, t.DecodeAccessKey)
+
+	// executing for each target in task
+	for _, tar := range t.Targets {
+		// getting new publisher
```
Comment (upendrasahu) [inline internal/task/task.go:189]: Use concurrent goroutines.

File: internal/configurator/configurator.go [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+func createTaskParam(auditJob input.AuditJob, targets []input.Target, decodedKey string) task.TaskParams {
+	var taskParam task.TaskParams
+	taskParam.Config = auditJob
+	taskParam.Targets = targets
+	taskParam.DecodeAccessKey = decodedKey
+	return taskParam
+}
+
+// convertIntervalToDuration converts the scheduling interval as provided in config.yaml to golang's duration
+func convertIntervalToDuration(interval string) time.Duration {
```
[the remark this answers is not in the document]
Comment (Nikhil-dot-Kumar) [reply internal/configurator/configurator.go:88]: ok

File: internal/task/task.go (hunk 6 above)
Comment (Nikhil-dot-Kumar) [reply internal/task/task.go:105]: ok

File: internal/task/task.go (hunk 7 above)
Comment (Nikhil-dot-Kumar) [reply internal/task/task.go:189]: ok will be adding concurrent go routines

File: internal/task/task.go [hunk 9]
```diff
@@ -0,0 +1,314 @@
...
+
+	// LastIssueTime represents the last issue time for the audit job.
+	LastIssueTime time.Time
+}
+
+func init() {
+	log = logger.GetLogger()
+
+	// checking if task stats file is present or not
+	_, err := os.Stat(taskStatsFile)
```
Comment (upendrasahu) [inline internal/task/task.go:88]: Log the error for debugging

File: internal/task/task.go [hunk 10]
```diff
@@ -0,0 +1,314 @@
...
+func SaveTaskStatsPeriodic(ctx context.Context) {
+	// 30 seconds periodic
+	ticker := time.NewTicker(30 * time.Second)
+	for {
+		select {
+		case <-ticker.C:
+			taskStatsMutex.Lock()
+			tBytes, err := json.Marshal(TaskStatsMap)
+			if err != nil {
+				log.Errorf("error[%v] in marshalling TaskStatsMap for checkpointing", err)
```
Comment (upendrasahu) [inline internal/task/task.go:112]: logging same error twice?

File: internal/task/task.go [hunk 11]
```diff
@@ -0,0 +1,314 @@
...
+				log.Error(err)
+			}
+			err = os.WriteFile(taskStatsFile, tBytes, 0644)
+			if err != nil {
+				log.Errorf("error[%v] in writing taskStatsFile file", err)
+			}
+			taskStatsMutex.Unlock()
+		case <-ctx.Done():
+			fmt.Println("stopping periodic save task stats routine")
```
Comment (upendrasahu) [inline internal/task/task.go:121]: instead of Println, it should be log

**Review by upendrasahu — APPROVED** [review]

**Pull request merged**